src/openassetio-python/private/include/openassetio/private/python/exceptions.hpp

```diff
@@ -0,0 +1,362 @@
+    static_assert(!std::is_base_of_v<CppExceptionsAndPyClassNames::Exceptions<J>,
+                                     CppExceptionsAndPyClassNames::Exceptions<I>>,
+                  "Base classes must come after subclasses in kCppExceptionsAndPyClassNames");
+  }
+
+  /**
+   * Convenience for a compile-time collection of indices in the list
+   * that come before the class under test.
+   */
+  static constexpr auto kPreviousIndices = std::make_index_sequence<I>{};
+};
+
+/**
+ * Assert that a set of classes (looked up by index) in the
+ * CppExceptionsAndPyClassNames list does not have any base classes
+ * appearing before them in the list.
+ *
+ * @tparam I Indices of classes to assert have no base classes at lower
+ * indices in the list.
+ */
+template <std::size_t... I>
+constexpr bool assertClassesAreInHierarchyOrder(
+    [[maybe_unused]] std::index_sequence<I...> unused) {
+  (AssertClassHierarchyForClassAtIndex<I>::assertNoneOfGivenClassIndicesIsABase(
+       AssertClassHierarchyForClassAtIndex<I>::kPreviousIndices),
+   ...);
+  return true;
+}
+/**
+ * Execute compile-time assertions that no class in the
+ * CppExceptionsAndPyClassNames list has a base class at a lower index
+ * in the list.
+ *
+ * The resulting value is unused - it's just kludge to make the
+ * compiler make the attempt and so fail on any violated static_asserts.
+ */
+[[maybe_unused]] constexpr bool kIsClassListOrderValid =
+    assertClassesAreInHierarchyOrder(CppExceptionsAndPyClassNames::kIndices);
+
+/// List of all OpenAssetIO-specific C++ exceptions.
+using openassetio::errors::AllExceptions;
+
+/**
+ * Assert that an exception in the AllExceptions list (looked up by
+ * index) is found in the given subset of the
+ * CppExceptionsAndPyClassNames list (looked up by indices).
+ *
+ * In practice the subset should be the full set.
+ *
+ * @tparam I Index of exception in AllExceptions.
+ * @tparam J Indices of exceptions in CppExceptionsAndPyClassNames.
+ */
+template <std::size_t I, std::size_t... J>
+constexpr void assertExceptionIsInList([[maybe_unused]] std::index_sequence<J...> unused) {
+  static_assert((std::is_same_v<std::tuple_element_t<I, AllExceptions>,
+                                CppExceptionsAndPyClassNames::Exceptions<J>> ||
+                 ...),
+                "Exception is missing from kCppExceptionsAndPyClassNames");
+}
+
+/**
+ * Assert that a subset of exceptions in the AllExceptions list (looked
+ * up by indices) are all found in the CppExceptionsAndPyClassNames
+ * list.
+ *
+ * In practice the subset should be the full set.
+ *
+ * @tparam I Indices of exceptions in AllExceptions.
+ */
+template <std::size_t... I>
+constexpr bool assertAllExceptionsAreInList([[maybe_unused]] std::index_sequence<I...> unused) {
+  (assertExceptionIsInList<I>(CppExceptionsAndPyClassNames::kIndices), ...);
+  return true;
+}
+/**
+ * Assert that all exceptions in the AllExceptions list are also found
```

Comment:
[Consider][minor] For a new reader, it might be nice to differentiate that the above are the mechanism by which the last actually does the assertion.

```suggestion
 * Execute compile-time assertions that all exceptions in the AllExceptions list are also found
```

Reply:
Updated docstring as suggested in 960a54e4